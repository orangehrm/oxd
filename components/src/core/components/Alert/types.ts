/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export const TYPE_DEFAULT = 'default';
export const TYPE_SUCCESS = 'success';
export const TYPE_WARN = 'warn';
export const TYPE_ERROR = 'error';
export const TYPE_INFO = 'info';

export const TYPES = [
  TYPE_DEFAULT,
  TYPE_SUCCESS,
  TYPE_WARN,
  TYPE_ERROR,
  TYPE_INFO,
];

export const ICON_MAP = {
  [TYPE_DEFAULT]: 'star',
  [TYPE_SUCCESS]: 'check2',
  [TYPE_WARN]: 'exclamation-triangle',
  [TYPE_ERROR]: 'exclamation-circle',
  [TYPE_INFO]: 'info-circle',
};

export type AlertType =
  | typeof TYPE_DEFAULT
  | typeof TYPE_SUCCESS
  | typeof TYPE_WARN
  | typeof TYPE_ERROR
  | typeof TYPE_INFO;
