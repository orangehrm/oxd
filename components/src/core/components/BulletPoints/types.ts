export interface BulletPointListItem {
    id: string;
    styles?: object;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
  }