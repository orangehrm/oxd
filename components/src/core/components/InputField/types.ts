export const COMPONENT_INPUT = 'oxd-input';
export const COMPONENT_FILE_INPUT = 'oxd-file-input';
export const COMPONENT_TEXTAREA = 'oxd-textarea';
export const COMPONENT_DROPDOWN_INPUT = 'oxd-dropdown-input';
export const COMPONENT_PASSWORD_INPUT = 'oxd-password-input';
export const COMPONENT_CHECKBOX_INPUT = 'oxd-checkbox-input';
export const COMPONENT_SWITCH_INPUT = 'oxd-switch-input';
export const COMPONENT_RADIO_INPUT = 'oxd-radio-input';
export const COMPONENT_DATE_INPUT = 'oxd-date-input';
export const COMPONENT_AUTOCOMPLETE_INPUT = 'oxd-autocomplete-input';
export const COMPONENT_SELECT_INPUT = 'oxd-select-input';
export const COMPONENT_MULTISELECT_INPUT = 'oxd-multiselect-input';
export const COMPONENT_TIME_INPUT = 'oxd-time-input';

export const TYPE_INPUT = 'input';
export const TYPE_FILE_INPUT = 'file';
export const TYPE_TEXTAREA = 'textarea';
export const TYPE_DROPDOWN = 'dropdown';
export const TYPE_PASSWORD = 'password';
export const TYPE_CHECKBOX = 'checkbox';
export const TYPE_SWITCH = 'switch';
export const TYPE_RADIO = 'radio';
export const TYPE_DATE = 'date';
export const TYPE_AUTOCOMPLETE = 'autocomplete';
export const TYPE_SELECT = 'select';
export const TYPE_MULTISELECT = 'multiselect';
export const TYPE_TIME = 'time';

export const TYPES = [
  TYPE_INPUT,
  TYPE_FILE_INPUT,
  TYPE_TEXTAREA,
  TYPE_DROPDOWN,
  TYPE_PASSWORD,
  TYPE_CHECKBOX,
  TYPE_SWITCH,
  TYPE_RADIO,
  TYPE_DATE,
  TYPE_AUTOCOMPLETE,
  TYPE_SELECT,
  TYPE_MULTISELECT,
  TYPE_TIME,
];

export const TYPE_MAP: TypeMap<Components> = {
  [TYPE_INPUT]: COMPONENT_INPUT,
  [TYPE_FILE_INPUT]: COMPONENT_FILE_INPUT,
  [TYPE_TEXTAREA]: COMPONENT_TEXTAREA,
  [TYPE_DROPDOWN]: COMPONENT_DROPDOWN_INPUT,
  [TYPE_PASSWORD]: COMPONENT_PASSWORD_INPUT,
  [TYPE_CHECKBOX]: COMPONENT_CHECKBOX_INPUT,
  [TYPE_SWITCH]: COMPONENT_SWITCH_INPUT,
  [TYPE_RADIO]: COMPONENT_RADIO_INPUT,
  [TYPE_DATE]: COMPONENT_DATE_INPUT,
  [TYPE_AUTOCOMPLETE]: COMPONENT_AUTOCOMPLETE_INPUT,
  [TYPE_SELECT]: COMPONENT_SELECT_INPUT,
  [TYPE_MULTISELECT]: COMPONENT_MULTISELECT_INPUT,
  [TYPE_TIME]: COMPONENT_TIME_INPUT,
};

export type Types =
  | typeof TYPE_INPUT
  | typeof TYPE_FILE_INPUT
  | typeof TYPE_TEXTAREA
  | typeof TYPE_DROPDOWN
  | typeof TYPE_PASSWORD
  | typeof TYPE_CHECKBOX
  | typeof TYPE_SWITCH
  | typeof TYPE_RADIO
  | typeof TYPE_DATE
  | typeof TYPE_AUTOCOMPLETE
  | typeof TYPE_SELECT
  | typeof TYPE_MULTISELECT
  | typeof TYPE_TIME;

export type Components =
  | typeof COMPONENT_INPUT
  | typeof COMPONENT_FILE_INPUT
  | typeof COMPONENT_TEXTAREA
  | typeof COMPONENT_DROPDOWN_INPUT
  | typeof COMPONENT_PASSWORD_INPUT
  | typeof COMPONENT_CHECKBOX_INPUT
  | typeof COMPONENT_SWITCH_INPUT
  | typeof COMPONENT_RADIO_INPUT
  | typeof COMPONENT_DATE_INPUT
  | typeof COMPONENT_AUTOCOMPLETE_INPUT
  | typeof COMPONENT_SELECT_INPUT
  | typeof COMPONENT_MULTISELECT_INPUT
  | typeof COMPONENT_TIME_INPUT;

export interface TypeMap<T> {
  [key: string]: T;
}
