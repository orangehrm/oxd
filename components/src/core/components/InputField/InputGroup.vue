<template>
  <div :class="rootClasses">
    <slot name="label">
      <div class="oxd-input-group__label-wrapper">
        <oxd-icon
          v-if="labelIcon"
          :name="labelIcon"
          class="oxd-input-group__label-icon"
        />
        <oxd-label v-if="label" :label="label" :class="labelClasses" />
      </div>
    </slot>
    <div>
      <slot></slot>
    </div>
    <slot name="message">
      <oxd-text v-if="message" :class="messageClasses" tag="span">
        {{ message }}
      </oxd-text>
    </slot>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Label from '@orangehrm/oxd/core/components/Label/Label.vue';
import Text from '@orangehrm/oxd/core/components/Text/Text.vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';

export default defineComponent({
  name: 'oxd-input-group',

  components: {
    'oxd-label': Label,
    'oxd-text': Text,
    'oxd-icon': Icon,
  },

  props: {
    label: {
      type: String,
    },
    labelIcon: {
      type: String,
    },
    message: {
      type: String,
    },
    classes: {
      type: Object,
      default() {
        return {
          label: {},
          message: {},
        };
      },
    },
  },

  computed: {
    rootClasses(): object {
      return {
        'oxd-input-group': true,
      };
    },
    labelClasses(): object {
      return {
        ...this.classes.label,
      };
    },
    messageClasses(): object {
      return {
        ...this.classes.message,
        'oxd-input-group__message': true,
      };
    },
  },
});
</script>

<style src="./input-group.scss" lang="scss" scoped></style>
