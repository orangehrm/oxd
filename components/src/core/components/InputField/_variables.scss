$oxd-input-field-error-color: $oxd-feedback-danger-color !default;
$oxd-input-field-bottom-space: 1rem !default;

$oxd-input-group-message-margin-top: 4px !default;
$oxd-input-group-message-font-size: 12px !default;
$oxd-input-group-message-font-weight: 600 !default;

$oxd-input-group-label-margin-bottom: 0.4rem !default;
$oxd-input-group-label-icon-margin-right: 4px !default;
