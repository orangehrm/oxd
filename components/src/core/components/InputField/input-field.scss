@import '../../../styles';
@import 'variables';

.oxd-input-field-error-message {
  color: $oxd-input-field-error-color !important;
}

.oxd-input-field-bottom-space {
  margin-bottom: $oxd-input-field-bottom-space;
}

.oxd-input-field-required:after {
  content: '*';
  color: $oxd-input-field-error-color;
}
