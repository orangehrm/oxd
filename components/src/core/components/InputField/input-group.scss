@import '../../../styles';
@import 'variables';

.oxd-input-group {
  display: flex;
  flex-direction: column;
  width: 100%;

  &__message {
    margin-top: $oxd-input-group-message-margin-top;
    font-size: $oxd-input-group-message-font-size;
    font-weight: $oxd-input-group-message-font-weight;
  }

  &__label-wrapper {
    display: flex;
    flex-direction: row;
    align-items: baseline;
    margin-bottom: $oxd-input-group-label-margin-bottom;
  }

  &__label-icon {
    margin-right: $oxd-input-group-label-icon-margin-right;
  }
}
