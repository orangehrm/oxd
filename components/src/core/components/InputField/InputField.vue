<template>
  <oxd-input-group
    :label="label"
    :labelIcon="labelIcon"
    :hint="hint"
    :hintPlacement="hintPlacement"
    :hintStyle="hintStyle"
    :id="id"
    :message="message"
    class="oxd-input-field-bottom-space"
    :classes="classes"
  >
    <component
      :is="component"
      v-bind="$attrs"
      :id="id"
      :disabled="disabled"
      :hasError="hasError"
      :modelValue="modelValue"
      @update:modelValue="onChange"
    >
      <template v-for="(_, name) in $slots" v-slot:[name]="slotData">
        <slot :name="name" v-bind="slotData" />
      </template>
    </component>
  </oxd-input-group>
</template>

<script lang="ts">
import {toRef, PropType, nextTick, defineComponent} from 'vue';
import InputGroup from '@orangehrm/oxd/core/components/InputField/InputGroup.vue';
import Input from '@orangehrm/oxd/core/components/Input/Input.vue';
import FileInput from '@orangehrm/oxd/core/components/Input/FileInput.vue';
import Textarea from '@orangehrm/oxd/core/components/Textarea/Textarea.vue';
import PasswordInput from '@orangehrm/oxd/core/components/Input/PasswordInput.vue';
import CheckboxInput from '@orangehrm/oxd/core/components/Input/CheckboxInput.vue';
import SwitchInput from '@orangehrm/oxd/core/components/Input/SwitchInput.vue';
import RadioInput from '@orangehrm/oxd/core/components/Input/RadioInput.vue';
import DateInput from '@orangehrm/oxd/core/components/Input/DateInput.vue';
import AutocompleteInput from '@orangehrm/oxd/core/components/Input/Autocomplete/AutocompleteInput.vue';
import QuickSearchInput from '@orangehrm/oxd/core/components/Input/Autocomplete/QuickSearchInput.vue';
import SelectInput from '@orangehrm/oxd/core/components/Input/Select/SelectInput.vue';
import MultiSelectInput from '@orangehrm/oxd/core/components/Input/MultiSelect/MultiSelectInput.vue';
import TimeInput from '@orangehrm/oxd/core/components/Input/Time/TimeInput.vue';
import Text from '@orangehrm/oxd/core/components/Text/Text.vue';
import InfoBox from '@orangehrm/oxd/core/components/InfoBox/InfoBox.vue';
import Comments from '@orangehrm/oxd/core/components/Comments/Comments.vue';
import TinyMce from '@orangehrm/oxd/core/components/TinyMce/TinyMce.vue';
import {
  Types,
  Components,
  TYPES,
  TYPE_INPUT,
  TYPE_MAP,
  HINT_PLACEMENT_TOP,
} from './types';
import useField from '../../../composables/useField';
import translateMixin from '../../../mixins/translate';
import CheckboxGroup from '@orangehrm/oxd/core/components/Input/CheckboxGroup.vue';
import RadioPillGroup from '@orangehrm/oxd/core/components/Input/RadioPills/RadioPillGroup.vue';
import TreeSelectInput from '@orangehrm/oxd/core/components/Input/TreeSelect/TreeSelect.vue';
import RadioGroup from '@orangehrm/oxd/core/components/Input/RadioGroup.vue';

export default defineComponent({
  name: 'oxd-input-field',
  inheritAttrs: false,

  components: {
    'oxd-input-group': InputGroup,
    'oxd-input': Input,
    'oxd-file-input': FileInput,
    'oxd-textarea': Textarea,
    'oxd-password-input': PasswordInput,
    'oxd-checkbox-input': CheckboxInput,
    'oxd-switch-input': SwitchInput,
    'oxd-radio-input': RadioInput,
    'oxd-date-input': DateInput,
    'oxd-autocomplete-input': AutocompleteInput,
    'oxd-select-input': SelectInput,
    'oxd-multiselect-input': MultiSelectInput,
    'oxd-time-input': TimeInput,
    'oxd-text': Text,
    'oxd-checkboxgroup-input': CheckboxGroup,
    'oxd-infobox': InfoBox,
    'oxd-quicksearch-input': QuickSearchInput,
    'oxd-comments': Comments,
    'oxd-radio-pill-group': RadioPillGroup,
    'oxd-tree-select-input': TreeSelectInput,
    'oxd-radiogroup-input': RadioGroup,
    'oxd-tinymce': TinyMce,
  },

  mixins: [translateMixin],
  emits: ['update:modelValue'],

  props: {
    modelValue: {},
    label: {
      type: String,
    },
    labelIcon: {
      type: String,
    },
    required: {
      type: Boolean,
      default: false,
    },
    id: {
      type: String,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String,
      default: TYPE_INPUT,
      validator: (value: Types) => {
        return TYPES.indexOf(value) !== -1;
      },
    },
    errors: {
      type: String,
      default: TYPE_INPUT,
      validator: (value: Types) => {
        return TYPES.indexOf(value) !== -1;
      },
    },
    rules: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      type: Array as PropType<any>,
      default: () => [],
    },
    hint: {
      type: String,
      default: null,
    },
    hintPlacement: {
      type: String,
      default: HINT_PLACEMENT_TOP,
    },
    hintStyle: {
      type: Object,
      default: () => ({}),
    },
    dirty: {
      type: Boolean,
      default: false,
    },
  },

  setup(props, context) {
    const modelValue = toRef(props, 'modelValue');
    const rules = toRef(props, 'rules');
    const isDisabled = toRef(props, 'disabled');
    const isDirty: boolean =
      props.dirty || Boolean(props.type === 'date' && modelValue.value);

    const initialValue = modelValue.value;

    const onReset = async () => {
      context.emit('update:modelValue', initialValue);
      await nextTick();
    };

    const {hasError, message, startWatcher, dirty} = useField({
      fieldLabel: props.label ? props.label : '',
      rules,
      modelValue,
      isDisabled,
      isDirty,
      onReset,
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const onChange = ($event: any) => {
      if (!dirty.value) {
        dirty.value = true;
        startWatcher();
      }
      context.emit('update:modelValue', $event);
    };

    if (isDirty) startWatcher();

    return {
      message,
      hasError,
      onChange,
    };
  },

  computed: {
    classes(): object {
      return {
        label: {
          'oxd-input-field-required': this.required,
        },
        message: {
          'oxd-input-field-error-message': this.hasError,
        },
      };
    },
    component(): Components {
      return TYPE_MAP[this.type as Types];
    },
  },
});
</script>

<style src="./input-field.scss" lang="scss"></style>
