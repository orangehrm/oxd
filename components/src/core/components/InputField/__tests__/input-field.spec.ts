/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount} from '@vue/test-utils';
import InputField from '@orangehrm/oxd/core/components/InputField/InputField.vue';
import {FormAPI, formKey} from '@orangehrm/oxd/composables/types';

describe('InputField.vue', () => {
  const mockFormAPI: FormAPI = {
    searchErrors: jest.fn(() => []),
    purgeErrors: jest.fn(),
    addError: jest.fn(),
    registerField: jest.fn(),
    unregisterField: jest.fn(),
  };
  it('renders OXD InputField', () => {
    const wrapper = mount(InputField, {
      props: {},
      global: {
        provide: {
          [formKey as symbol]: mockFormAPI,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD InputField type `input`', () => {
    const wrapper = mount(InputField, {
      props: {type: 'input'},
      global: {
        provide: {
          [formKey as symbol]: mockFormAPI,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD InputField type `file`', () => {
    const wrapper = mount(InputField, {
      props: {type: 'file'},
      global: {
        provide: {
          [formKey as symbol]: mockFormAPI,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD InputField type `file` with button', () => {
    const wrapper = mount(InputField, {
      props: {type: 'file', buttonLabel: 'Browse'},
      global: {
        provide: {
          [formKey as symbol]: mockFormAPI,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders OXD InputField type `textarea`', () => {
    const wrapper = mount(InputField, {
      props: {type: 'textarea'},
      global: {
        provide: {
          [formKey as symbol]: mockFormAPI,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
