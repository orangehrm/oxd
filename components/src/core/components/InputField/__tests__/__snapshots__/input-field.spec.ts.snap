// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField.vue renders OXD InputField 1`] = `
<div class="oxd-input-group oxd-input-field-bottom-space">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div><input class="oxd-input oxd-input--active"></div>
  <!--v-if-->
</div>
`;

exports[`InputField.vue renders OXD InputField type \`file\` 1`] = `
<div class="oxd-input-group oxd-input-field-bottom-space">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div><input type="file" class="oxd-file-input">
    <div class="oxd-input oxd-file-div oxd-input--active">
      <!--v-if-->
      <div class="oxd-file-input-div"></div>
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`InputField.vue renders OXD InputField type \`file\` with button 1`] = `
<div class="oxd-input-group oxd-input-field-bottom-space">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div><input type="file" class="oxd-file-input">
    <div class="oxd-input oxd-file-div oxd-input--active">
      <div class="oxd-file-button">Browse</div>
      <div class="oxd-file-input-div"></div>
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`InputField.vue renders OXD InputField type \`input\` 1`] = `
<div class="oxd-input-group oxd-input-field-bottom-space">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div><input class="oxd-input oxd-input--active"></div>
  <!--v-if-->
</div>
`;

exports[`InputField.vue renders OXD InputField type \`textarea\` 1`] = `
<div class="oxd-input-group oxd-input-field-bottom-space">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div><textarea class="oxd-textarea oxd-textarea--active oxd-textarea--resize-vertical"></textarea></div>
  <!--v-if-->
</div>
`;
